src/ReverseProxy/WebSocketsTelemetry/WebSocketsTelemetryStream.cs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Yarp.ReverseProxy.Forwarder;
+using Yarp.ReverseProxy.Utilities;
+
+namespace Yarp.ReverseProxy.WebSocketsTelemetry
+{
+    internal sealed class WebSocketsTelemetryStream : DelegatingStream
+    {
+        private WebSocketsParser _readParser, _writeParser;
+
+        public DateTime EstablishedTime { get; }
+        public long MessagesRead => _readParser.MessageCount;
+        public long MessagesWritten => _writeParser.MessageCount;
+
+        public WebSocketsTelemetryStream(Stream innerStream)
+            : base(innerStream)
+        {
+            EstablishedTime = DateTime.UtcNow;
+            _readParser = new WebSocketsParser(isServer: true);
+            _writeParser = new WebSocketsParser(isServer: false);
+        }
+
+        public WebSocketCloseReason GetCloseReason(HttpContext context)
+        {
+            var clientCloseTime = _readParser.CloseTime;
+            var serverCloseTime = _writeParser.CloseTime;
+
+            if (clientCloseTime.HasValue && (!serverCloseTime.HasValue || serverCloseTime.Value > clientCloseTime.Value))
```

Comment:
The close is only graceful if both sides acknowledge it. Otherwise you can attribute the reason as a disconnect by the unacknowledged side.
```suggestion
            if (clientCloseTime.HasValue && serverCloseTime.HasValue && serverCloseTime.Value > clientCloseTime.Value)
```

Reply:
So something like
```c#
if (clientSentClose && serverSentClose)
{
    return clientCloseTime < serverCloseTime ? ClientGraceful : ServerGraceful;
}
else
{
    Disconnect
}
```
?

Reply:
```c#
if (clientSentClose && serverSentClose)
{
    return clientCloseTime < serverCloseTime ? ClientGraceful : ServerGraceful;
}
else if (clientSentClose)
{
    ServerDisconnect
}
else if (serverSentClose)
{
   ClientDisconnect
}
else // check forwarder errors
```

Reply:
If the client sends a Close message and immediately closes the connection, we would be reporting that as a ServerDisconnect (and vice-versa).

Would that be the desired behavior?

Reply:
Interesting... how would you detect that? Would the read return 0? Writes might fail, depending on the timing.

Reply:
I changed the logic so that it is now
```c#
if (clientSentClose && serverSentClose)
{
    return clientCloseTime < serverCloseTime ? ClientGraceful : ServerGraceful;
}

// One side sent a WebSocket close, but we never saw a response from the other side
// It is possible an error occurred, but we saw a graceful close first, so that is the intiator
if (clientSentClose)
{
    return ClientGraceful;
}
if (serverSentClose)
{
    return ServerGraceful;
}

return forwarderError ...
```

While "graceful" doesn't necessarily mean "both sides sent close", I believe reporting what likely initiated the close is more useful as the `WebSocketCloseReason`.
Either side may still perform protocol violations, but it's not our aim to catch that.